components/brave_wallet/browser/eth_json_rpc_controller.cc

```diff
@@ -133,28 +130,43 @@ void EthJsonRpcController::SetNetwork(Network network) {
                          subdomain.c_str(), GetInfuraProjectID().c_str());
   network_url_ = GURL(spec);
 
-  observers_->Notify(FROM_HERE,
-                     &BraveWalletProviderEventsObserver::ChainChangedEvent,
-                     GetChainIDFromNetwork(network_));
+  for (const auto& observer : observers_) {
+    observer->ChainChangedEvent(GetChainIdFromNetwork(network_));
+  }
+}
+
+void EthJsonRpcController::GetChainId(
+    mojom::EthJsonRpcController::GetChainIdCallback callback) {
+  std::move(callback).Run(GetChainIdFromNetwork(network_));
+}
+
+void EthJsonRpcController::GetBlockTrackerUrl(
+    mojom::EthJsonRpcController::GetBlockTrackerUrlCallback callback) {
+  std::move(callback).Run(GetBlockTrackerUrlFromNetwork(network_).spec());
+}
+
+void EthJsonRpcController::GetNetworkUrl(
+    mojom::EthJsonRpcController::GetNetworkUrlCallback callback) {
+  std::move(callback).Run(network_url_.spec());
 }
 
 void EthJsonRpcController::SetCustomNetwork(const GURL& network_url) {
-  network_ = Network::kCustom;
+  network_ = brave_wallet::mojom::Network::Custom;
```

Comment:
having `SetNetwork` and `SetCustomNetwork` seems problematic because of https://github.com/brave/brave-core/pull/9564/files#r677616338
Seems like it would be better for `SetNetwork` to take an optional url param that is required for `Custom`. This also means that `SetNetwork` should take a callback to notify the caller of the error if they don't supply the url

Reply:
that seems a bit like overkill to have a callback. If a user sets custom directly then it just won't work correctly. We could instead simply DCHECK for unexpected API usage. A user could call your suggested `SetNetwork` incorrectly by passing a URL when mainnet is specified.

Reply:
per discussion on Slack we had a compromise to call NOTREACHED() for the invalid API usage.